design-docs/001-virtual-chunk-improvements.md

```diff
@@ -0,0 +1,381 @@
+# Improvements to virtual chunks handling
+
+## What we need
+
+1. Virtual chunks can be stored in all supported object stores. The same repo can have virtual chunks hosted in different object stores.
+2. Multiple credentials sets for virtual chunk resolution, accessing some chunks may require a different set than others.
+3. A mechanism to know all the "dependencies" of a repo. That is, all the "places" where its virtual chunks are hosted. This is important for users to know if they can delete old data, or if there are virtual datasets that depend on it.
+4. If the file hosting a virtual chunk changed after the chunk ref was written, we want to fail with a runtime error instead of serving corrupted data. This is because offset and length are usually not maintained during updates.
+
+## Discussion
+
+### Current virtual chunk URL styles
+
+* The open source stack seating on top of Icechunk (VirtualiZarr, etc.) uses URLs to identify virtual chunk locations.
+* One common case is many virtual chunks from a few objects.
+  * Example: `s3://some-bucket/some-prefix/some-file.nc` is pointed to by 10k chunks.
+  * Having between 10 and 10k chunks from the same object is not unusual
+* Other common use case for virtual zarr datasets: each chunk comes from a different object. This can happen, for example, with a virtual zarr dataset.
```

Comment:
```suggestion
* The other common use case for virtual zarr datasets: each chunk comes from a different object. This can happen, for example, with a virtual zarr dataset.
```